config/src/credentials.rs

```diff
@@ -37,26 +41,152 @@ impl Credentials {
+
+        // decode json string to struct
+        let key_cloak_json: KeyCloakJson = serde_json::from_slice(&buffer).expect("decode slice to struct");
+
+        // first POST request for access token
+        let access_token = obtain_access_token(
+            &key_cloak_json.auth_server_url,
+            &key_cloak_json.realm,
+            &key_cloak_json.resource,
+            &key_cloak_json.credentials.secret,
+        )
+        .await
+        .expect("obtain access token");
 
-    pub fn get_authentication_token(&self) -> String {
-        self.token.clone()
+        // second POST request for rpt
+        let rpt = authorize(
+            &key_cloak_json.auth_server_url,
+            &key_cloak_json.realm,
+            &access_token,
+        )
+        .await
+        .expect("get rpt");
+        Credentials {
+            method: BEARER.to_owned(),
+            token: rpt,
+        }
     }
 
     pub fn get_request_metadata(&self) -> String {
         format!("{} {}", self.method, self.token)
     }
 }
 
+#[derive(Serialize, Deserialize)]
+struct KeyCloakJson {
+    realm: String,
+    #[serde(rename(deserialize = "bearer-only"))]
+    bearer_only: bool,
+    #[serde(rename(deserialize = "auth-server-url"))]
+    auth_server_url: String,
+    #[serde(rename(deserialize = "ssl-required"))]
+    ssl_required: String,
+    resource: String,
+    #[serde(rename(deserialize = "confidential-port"))]
+    confidential_port: i32,
+    #[serde(rename(deserialize = "public-client"))]
+    public_client: bool,
+    credentials: Credential,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Credential {
+    secret: String,
+}
+
+async fn obtain_access_token(
+    base_url: &str,
+    realm: &str,
+    client_id: &str,
+    client_secret: &str,
+) -> Result<String, reqwest::Error> {
+    let url = URL_TOKEN.replace("{realm-name}", realm);
+
+    let payload = serde_json::json!({
+        "client_id": client_id.to_owned(),
+        "client_secret": client_secret.to_owned(),
+        "grant_type": CLIENT_CREDENTIAL_GRANT_TYPE.to_owned(),
+    });
+
+    let path = base_url.to_owned() + &url.to_owned();
+
+    let mut header_map = HeaderMap::new();
+    header_map.insert(CONTENT_TYPE, "application/json".parse().unwrap());
+    let token = send_http_request(&path, payload, header_map).await?;
+    Ok(token.access_token)
+}
+
+async fn authorize(base_url: &str, realm: &str, token: &str) -> Result<String, reqwest::Error> {
+    let url = URL_TOKEN.replace("{realm-name}", realm);
+
+    let payload = serde_json::json!({
+        "grant_type": UMA_GRANT_TYPE.to_owned(),
+        "audience": AUDIENCE.to_owned(),
+    });
+
+    let path = base_url.to_owned() + &url.to_owned();
+
+    let mut header_map = HeaderMap::new();
+    let bearer = format!("{} {}", BEARER, token);
+    header_map.insert(AUTHORIZATION, bearer.parse().unwrap());
+    let rpt = send_http_request(&path, payload, header_map).await?;
+    Ok(rpt.access_token)
```

Comment:
I'm not sure if it's in the scope of this PR, but I just wanted to point out that in the response that contains the token in the "access_token" json field, there's also a "expires_in" field (in seconds).

This can be helpful for some soft caching of the rpt such that you can intelligently decide to not make a request for the rpt (not the access token) if you know you have one and it's far enough away from the expires in time.

(doesn't preclude from having to handle token expiration/errors coming back from the pravega controller when the token is bad; but it can optimize traffic to Keycloak.

Reply:
Thanks for the comment! 

My understanding is that once an instance of Rust client gets the rpt token, it will never ask for it again. Is that correct?
If that's the case, I think we don't need cache here because rust keycloak client will never be called twice.

I am guessing that the reason that java keycloak client has cache is perhaps Java keycloak client is a service and every Pravega client needs to call it?

Reply:
Just to clarify a couple things:

I don't know how long the Rust client lives; and if perhaps it's interaction with Pravega is a one shot deal.  But just in case it's not; all I meant is that the RPT token that gets a hold of will expire after 5 minutes.

The PravegaKeycloakCredentials is not a service per say; it's merely just a library, implementation of the Pravega client side's Credentials interface which gets loaded at runtime.   The reason why it caches the RPT for some time, is that it knows it might live in the context of a long running application where the Pravega Client remains loaded for a long time.  As such, it needs to be able to get new tokens when asked; and if possible be a bit smart about it a cache the said token.

I don't know if all of this translate to this use case and how the rust client is actually used.   Does it make sense?